### PR TITLE
[pigeon] Initial integration test setup

### PR DESCRIPTION
**Pull request by stuartmorgan**
This sets up initial proof-of-concept integration tests using the new shared native test harness:
- Integration tests on the Dart side for void->void and Everything->Everything calls.
- macOS implementations in the test plugin on the native side.
- A new test target in the test script to drive them via `flutter test`.
- A minimal change to the example app so that `flutter run`-ing it will test that the void->void call is wired up.

Since this simple initial test hit
https://github.com/flutter/flutter/issues/111083, which caused the test to fail, this includes a fix for that.

Short-term future work (by me):
- Add integration test native setup and script targets for the other generators. This includes one just to keep the initial review scope smaller.
- Update https://github.com/flutter/packages/pull/2816 to include the integration test since it's still blocked until I can address the CI issues.

Medium-term future work (not all by me):
- Remove the legacy iOS e2e test scaffold that is currently disabled.
- Add significantly more integration test coverage (likely including https://github.com/flutter/flutter/issues/115168 to reduce redundant API setup), including Flutter API integration tests rather than just host API tests.

Part of https://github.com/flutter/flutter/issues/111505
Fixes https://github.com/flutter/flutter/issues/111083

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/pigeon/lib/swift_generator.dart [hunk 1]
```diff
@@ -232,7 +232,7 @@ void _writeHostApi(Indent indent, Api api, Root root) {
               indent.write('$call ');
               if (method.returnType.isVoid) {
                 indent.scoped('{', '}', () {
-                  indent.writeln('reply(nil)');
```
Comment (gaaclarke) [inline packages/pigeon/lib/swift_generator.dart:235]: I thought we fixed this, thanks.

File: packages/pigeon/platform_tests/test_plugin/example/lib/main.dart [hunk 2]
```diff
@@ -21,6 +22,8 @@ class MyApp extends StatefulWidget {
 class _MyAppState extends State<MyApp> {
   // ignore: unused_field
   final TestPlugin _testPlugin = TestPlugin();
+  late AllVoidHostApi api;
```
Comment (gaaclarke) [inline packages/pigeon/platform_tests/test_plugin/example/lib/main.dart:25]: final

File: packages/pigeon/platform_tests/test_plugin/macos/Classes/TestPlugin.swift [hunk 3]
```diff
@@ -6,14 +6,34 @@ import Cocoa
...
  */
-public class TestPlugin: NSObject, FlutterPlugin {
+public class TestPlugin: NSObject, FlutterPlugin, AllVoidHostApi, HostEverything {
   public static func register(with registrar: FlutterPluginRegistrar) {
+    let plugin = TestPlugin()
+    AllVoidHostApiSetup.setUp(binaryMessenger: registrar.messenger, api: plugin)
+    HostEverythingSetup.setUp(binaryMessenger: registrar.messenger, api: plugin)
   }
 
   public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
```
Comment (gaaclarke) [inline packages/pigeon/platform_tests/test_plugin/macos/Classes/TestPlugin.swift:19]: Can you not pull this out?  It's a bit confusing to have.

File: packages/pigeon/tool/run_tests.dart [hunk 4]
```diff
@@ -61,9 +62,12 @@ const Map<String, _TestInfo> _tests = <String, _TestInfo>{
   'ios_swift_unittests': _TestInfo(
       function: _runIosSwiftUnitTests,
       description: 'Unit tests on generated Swift code.'),
-  'mac_swift_unittests': _TestInfo(
+  'macos_swift_unittests': _TestInfo(
       function: _runMacOSSwiftUnitTests,
       description: 'Unit tests on generated Swift code on macOS.'),
+  'macos_swift_integration_tests': _TestInfo(
```
Comment (gaaclarke) [inline packages/pigeon/tool/run_tests.dart:68]: These are called e2e tests above, we should standardize the name.

File: packages/pigeon/platform_tests/test_plugin/example/lib/main.dart (hunk 2 above)
Comment (gaaclarke) [inline packages/pigeon/platform_tests/test_plugin/example/lib/main.dart:25]: Shouldn't there be an EverythingApi too somewhere?

File: packages/pigeon/platform_tests/test_plugin/macos/Classes/TestPlugin.swift (hunk 3 above)
Comment (stuartmorgan) [reply packages/pigeon/platform_tests/test_plugin/macos/Classes/TestPlugin.swift]: I forgot it was marked as `@optional` in the protocol; removed.

File: packages/pigeon/platform_tests/test_plugin/example/lib/main.dart (hunk 2 above)
Comment (stuartmorgan) [reply packages/pigeon/platform_tests/test_plugin/example/lib/main.dart]: Done.

**Review by gaaclarke — APPROVED** [review]
lgtm!

**Pull request merged**